src/drivers/distance_sensor/tfmini/TFMINI.cpp

```diff
@@ -0,0 +1,266 @@
+		int termios_state{};
+
+		tcgetattr(_fd, &uart_config);
+
+		// clear ONLCR flag (which appends a CR for every LF)
+		uart_config.c_oflag &= ~ONLCR;
+
+		// set baud rate
+		if ((termios_state = cfsetispeed(&uart_config, speed)) < 0) {
+			PX4_ERR("CFG: %d ISPD", termios_state);
+			ret = -1;
+			break;
+		}
+
+		if ((termios_state = cfsetospeed(&uart_config, speed)) < 0) {
+			PX4_ERR("CFG: %d OSPD\n", termios_state);
+			ret = -1;
+			break;
+		}
+
+		if ((termios_state = tcsetattr(_fd, TCSANOW, &uart_config)) < 0) {
+			PX4_ERR("baud %d ATTR", termios_state);
+			ret = -1;
+			break;
+		}
+
+		uart_config.c_cflag |= (CLOCAL | CREAD);	// ignore modem controls
+		uart_config.c_cflag &= ~CSIZE;
+		uart_config.c_cflag |= CS8;			// 8-bit characters
+		uart_config.c_cflag &= ~PARENB;			// no parity bit
+		uart_config.c_cflag &= ~CSTOPB;			// only need 1 stop bit
+		uart_config.c_cflag &= ~CRTSCTS;		// no hardware flowcontrol
+
+		// setup for non-canonical mode
+		uart_config.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL | IXON);
+		uart_config.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
+		uart_config.c_oflag &= ~OPOST;
+
+		// fetch bytes as they become available
+		uart_config.c_cc[VMIN] = 1;
+		uart_config.c_cc[VTIME] = 1;
+
+		if (_fd < 0) {
+			PX4_ERR("FAIL: laser fd");
+			ret = -1;
+			break;
+		}
+	} while (0);
+
+	// close the fd
+	::close(_fd);
+	_fd = -1;
+
+	if (ret == PX4_OK) {
+		start();
+	}
+
+	return ret;
+}
+
+int
+TFMINI::collect()
+{
+	perf_begin(_sample_perf);
+
+	// clear buffer if last read was too long ago
+	int64_t read_elapsed = hrt_elapsed_time(&_last_read);
+
+	// the buffer for read chars is buflen minus null termination
+	char readbuf[sizeof(_linebuf)] {};
+	unsigned readlen = sizeof(readbuf) - 1;
+
+	int ret = 0;
+	float distance_m = -1.0f;
+
+	// Check the number of bytes available in the buffer
+	int bytes_available = 0;
+	::ioctl(_fd, FIONREAD, (unsigned long)&bytes_available);
+
+	if (!bytes_available) {
+		perf_end(_sample_perf);
+		return -EAGAIN;
+	}
+
+	// parse entire buffer
+	const hrt_abstime timestamp_sample = hrt_absolute_time();
+
+	int ret_parse = PX4_ERROR;
+
+	do {
+		// read from the sensor (uart buffer)
+		ret = ::read(_fd, &readbuf[0], readlen);
+
+		if (ret < 0) {
+			PX4_ERR("read err: %d", ret);
+			perf_count(_comms_errors);
+			perf_end(_sample_perf);
+
+			// only throw an error if we time out
+			if (read_elapsed > (kCONVERSIONINTERVAL * 2)) {
+				/* flush anything in RX buffer */
+				tcflush(_fd, TCIFLUSH);
+				return ret;
+
+			} else {
+				return -EAGAIN;
+			}
+		}
+
+		_last_read = hrt_absolute_time();
+
+		// parse buffer
+		for (int i = 0; i < ret; i++) {
+			ret_parse = tfmini_parse(readbuf[i], _linebuf, &_linebuf_index, &_parse_state, &distance_m);
+		}
+
+		// bytes left to parse
+		bytes_available -= ret;
+
+	} while (bytes_available > 0);
+
+	// no valid measurement after parsing buffer
+	if (distance_m < 0.0f) {
```

Comment:
@dagar this was doing the equivalent of the ret_parse you just added.

Reply:
Ah, thanks. I'll remove it.